keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+  where traffic should be forwarded, ex: updating the `externalTrafficPolicy`
+  field from `Cluster` to `Local` or vice versa. Such changes to the service
+  object should not end up re-configuring the load balancer, only change the
+  response returned by the service proxy to the load balancer's HC.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+what are all the fields on the service object which fulfill this statement?
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Avoiding re-configuration of the load balancer for fully deleted / newly added
+  cluster nodes.
+
+## Proposal
+
+Up until this point the KCCM has defined the following conditions for what is
+considered a "healthy" node:
+
+1. Is the node not tainted with the key: `ToBeDeletedByClusterAutoscaler`?
+2. Is the node condition `NodeReady` equal to `ConditionTrue`?
+
+This first predicate was introduced by the
+[Kube PR #105946](https://github.com/kubernetes/kubernetes/pull/105946) which
+stated:
+
+> If multiple...nodes are being deleted, the list of nodes passed to
+> cloud-provider code via UpdateLoadBalancer() will be stale, causing a few
+> failed updates before the backends stabilize.
+
+It seemed, from reading the above, that the goal was to ensure that we get a
+head start at deleting the node from the load balancer set before the underlying
+VM instance has been completely cleaned up by the autoscaler. The interpretation
+of the PR indicated that if the predicate is not used: we'll only get notified
+the moment the node is deleted in the API server, which might be after the VM
+instance is GC'ed, causing the update call on the load balancer to fail because
+the node cannot be found - and thus, possibly, stale references by the load
+balancer to nodes which don't exists anymore.
+
+This cannot be true today however. This is easily logically verifiable by simply
+looking at the generic cloud-provider API for load balancer operations that the
+KCCM can invoke: the only functions which ends up getting called when a node is
+deleted is: `EnsureLoadBalancer`/`UpdateLoadBalancer`. But the list of nodes
+provided as an argument to both (`nodes`) should contain __all the remaining
+ cluster nodes__. It is therefor impossible that deleting a Node from the LB set
+__after__ the VM has been deleted, should cause any problem. The cloud-provider
+API can only be invoked using all the remaining and known Nodes/VMs. Either the
+reasons for adding this condition in the first place was misunderstood, or they
+were wrong or they might have been right __at the time__, but with the current
+code it certainly doesn't make sense to keep.
+
+The second predicate has no direct correlation to networking, but __could__,
+__possibly__, __maybe__ indicate a general problem with the node should the
+Kubelet be un-healthy in certain specific cases (for ex: if it's resource
+starved, which could be impacting other system processes as well, and
+subsequently networking related processes).
```

Comment:
Another issue could be, depending on why the node was marked as unready, that the kube-proxy has a stale list of endpoints. A couple edge cases we would have to think about if we were to loosen this predicate: 
* What happens if the pod goes away?
* What happens if a new pod with an old ip is running?

Reply:
If we decide we won't loosen it because we are afraid of these failure modes. Then we should, at the very least, define a node condition which accurately reflects the "node's networking state". The only explicit "network problem" that might transition the node into this state is if the CNI plugin didn't write its config file to disk...this is not good enough, and we now remove it from the LB set because, at best, CPU / Mem / disk pressure might be impacting networking too.